src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -526,12 +503,22 @@ public static void sleep(long millis, int nanos) throws InterruptedException {
             return;
         }
 
-        if (nanos > 0 && millis < Long.MAX_VALUE) {
-            millis++;
+        if (ThreadSleepEvent.isTurnedOn()) {
+            ThreadSleepEvent event = new ThreadSleepEvent();
+            try {
+                event.time = MILLISECONDS.toNanos(millis) + nanos;
```

Comment:
This can overflow when millis is at or close to Long.MAX_VALUE.

Reply:
> Yes, let me fix that. `TimeUnit.toNanos` handles it well itself, it seems.

This code is refactored in PR 13203 so we'll have to merge at some point.

Reply:
Merged!